src/app/Archives/ArchiveUploadModal.tsx

```diff
@@ -122,40 +136,126 @@ export const ArchiveUploadModal: React.FunctionComponent<ArchiveUploadModalProps
     }
   }, [uploading, setShowCancelPrompt, reset, props.onClose]);
 
+  const getProgressUpdateCallback = React.useCallback(
+    (filename: string) => {
+      return (process: string | number) => {
+        setFileUploads((old) => {
+          const match = old.find((f) => f.file.name === filename);
+          if (match) {
+            return [
+              ...old.filter((f) => f.file.name !== filename),
+              {
+                ...match,
+                progress: {
+                  progressValue: process,
+                  progressVariant: process == 100 ? 'success' : undefined,
+                },
+              } as FUpload,
+            ];
+          }
+          return old;
+        });
+      };
+    },
+    [setFileUploads]
+  );
```

Comment:
Is there any refactoring that can be done to extract these callbacks so they're reusable across components? It seems like they're basically just copied and pasted between here and the `AgentProbleTemplates` for example. It'd be nice to have a single reused implementation.